### PR TITLE
Replace graph

### PR DESCRIPTION
**Pull request by tfinnell**
The main change here is `Facebook.Graph` is replaced with `Facebook.GraphAPI` & `Facebook.GraphVideoAPI`, the HTTP response formatting functions were moved into a separate module, `ResponseFormatter`. One thing to consider is combining and moving the http request +  formatter calls into the existing `Facebook.Graph`, but I do like having the HTTP library exposed in the main library opposed to in another wrapper as we get a bit more flexibility in configuring requests without having to modify a wrapper module.

I ended up updating a bit of documentation as well since I was basically touching every function as I was doing this. Added some `@typedocs` and provided a few links here and there. Also added a blurb about contributing to the README.

When we bump to v2.11 of the graph API we may be able to ditch the additional wrapper around the video api as I believe it's all now under the single `graph_url`. Things still seemed to work as they are on v2.11 when I tested, though.

Last thing, looks like #48 is stagnant, so, if there's no feedback that results in major changes to this PR, I'll start looking into implementing the `chat` endpoints on top of this. Regardless, appreciate any feedback guys!

### CODE REVIEW

**Comment by paulruescher** [comment]
Nice 👍 

I just took a quick look and haven't done a deep dive, but one thing sticks out: do we even need `hackney` anymore?

**Comment by tfinnell** [comment]
Ya, you're right. HTTPoison is pulling it in anyway.

File: README.md [hunk 1]
```diff
@@ -37,23 +37,14 @@ Then try some API calls:
...
 ```
 
-## API Documentation
-
-See [API Documentation for facebook.ex](http://hexdocs.pm/facebook/).
+## Contributing
+We encourage contribution from anyone! If you've got an improvement to the documentation or feature you've implemented, please open a [pull request](https://github.com/mweibel/facebook.ex/pulls).
+This project uses [credo](https://github.com/rrrene/credo) for code analysis. Running `mix credo` will give you a nice output which will tell you if any of the changes you've made aren't consistent with the rest of our codebase.
 
+The Facebook Graph API is fairly large and as such we're not using every facet of it, so if you're not seeing an edge that is handled, please report an [issue](https://github.com/mweibel/facebook.ex/issues).
```
Comment (mweibel) [inline README.md]: report an issue or open a PR to add it?

File: README.md [hunk 2]
```diff
@@ -37,23 +37,14 @@ Then try some API calls:
...
+iex(2)> Facebook.object_count(:likes, "262588213843476_801732539929038", "ACCESSTOKEN")
+{:ok, 48} # <--- that's the return value
 ```
 
-## API Documentation
-
-See [API Documentation for facebook.ex](http://hexdocs.pm/facebook/).
+## Contributing
+We encourage contribution from anyone! If you've got an improvement to the documentation or feature you've implemented, please open a [pull request](https://github.com/mweibel/facebook.ex/pulls).
+This project uses [credo](https://github.com/rrrene/credo) for code analysis. Running `mix credo` will give you a nice output which will tell you if any of the changes you've made aren't consistent with the rest of our codebase.
```
Comment (mweibel) [inline README.md]: great!

**Comment by mweibel** [comment]
Just had a brief look but so far this looks amazing! Thanks for that!

File: lib/facebook/graph_api.ex [hunk 3]
```diff
@@ -0,0 +1,14 @@
+defmodule Facebook.GraphAPI do
+  @moduledoc false
+
+  use HTTPoison.Base
+
+  alias Facebook.Config
+
+  def process_url(url), do: Config.graph_url <> url
```
Comment (paulruescher) [inline lib/facebook/graph_api.ex]: I have a pattern matching idea for how to combine this with the `GraphVideoApi` module, I just need a sec to test it. Essentially I think we can pattern match off a binary passed in

File: test/graph_mock.ex [hunk 4]
```diff
@@ -101,4 +101,18 @@ defmodule Facebook.GraphMock do
...
   end
+
+  def mock_options(body_function) do
+    [
+      request: request(),
+      body: body_function,
+    ]
+  end
+
+  defp request do
```
Comment (paulruescher) [inline test/graph_mock.ex]: so much cleaner than what I had 👍

File: test/facebook_test.exs [hunk 5]
```diff
@@ -21,25 +23,19 @@ defmodule FacebookTest do
 
   describe "me" do
     test "success", %{id: id, access_token: access_token} do
-      with_mock :hackney, [
-        request: fn(_method, _url, _headers, _payload, _options) ->
-          {:ok, nil, nil, nil}
-        end,
-        body: fn(_) -> Facebook.GraphMock.me(:success) end
-      ] do
+      with_mock :hackney, GraphMock.mock_options(
```
Comment (mweibel) [inline test/facebook_test.exs]: :hackney mock is probably not needed anymore, or is it?

**Comment by mweibel** [comment]
Went a bit more through it and that all seems ok. A bit confused about the hackney mocks there. Is it because HTTPoison uses hackney internally? 
Anyway, let me know when you consider this done and I'm more than happy to merge!
If you'd like I'll also make you an owner of the package on hex so you could publish new versions too.

**Comment by paulruescher** [comment]
Oops, looks I forgot to submit my review last night 🤷‍♀️

**Comment by tfinnell** [comment]
Exactly! HTTPoison is just a wrapper for hackney (just as `Facebook.Graph` was), which is why this still works.

Now, we *could* be explicit and go the route of mocking the HTTPoison method functions, but I sorta like the consistency as is.

Since it does seem to raise some `wtf?` reactions, I've added some documentation to that module :smiley: 

Thanks for the feedback @paulruescher & @mweibel -- I think this is good to go.

@mweibel I'd be happy to help out with version publishing!

**Comment by paulruescher** [comment]
Awesome, I def agree with keeping it for now. I know we talked mocking briefly before, but I've been using this as a guide recently for structuring 3rd party API calls of my own http://blog.plataformatec.com.br/2015/10/mocks-and-explicit-contracts/

Seems to work quite well. Definitely future consideration tho 👍 

Awesome work 🎉

**Comment by tfinnell** [comment]
I read that same blog post as I was working on this as well :smile:
I keep bringing up this library, but, `exVCR` seems to address this mocking issue since the http library it uses is configurable. The reason I'm so keen on it is the amount of transparency it provides into the request/responses that an http library creates -- it becomes a super useful tool for debugging!

**Comment by mweibel** [comment]
Thanks!
Merged it. Can you tell me your hex.pm email address so I can add you? (You can also send it to my email address in my git commits if you don't wanna put it here)

**Pull request merged**